pip/index.py

```diff
@@ -486,9 +486,32 @@ def find_requirement(self, req, upgrade):
             c for c in all_candidates if str(c.version) in compatible_versions
         ]
 
+        yanked_candidates = [
+            c for c in applicable_candidates
+            if getattr(c.location, "yanked", False)
+        ]
+
         if applicable_candidates:
             best_candidate = max(applicable_candidates,
                                  key=self._candidate_sort_key)
+            # If we cannot find a non-yanked candidate,
+            # use the best one and print a warning about it.
+            # Otherwise, try to find another best candidate, ignoring
+            # all the yanked releases.
+            if best_candidate in yanked_candidates:
+                if set(applicable_candidates) <= set(yanked_candidates):
```

Comment:
Nitpick: `if best_candidate in yanked_candidates` is the same as `if getattr(best_candidate, 'yanked', False)`, isn't it?
It might be easier to find the `set` of `nonyanked_candidates` here, and reuse it below.

Reply:
I'm not sure what you meant by this so I've reversed the logic to have nonyanked_candidates which means that I need to loop through all applicable candidates only once which I think is the wanted outcome. See the latest fixup commit.

Reply:
I meant that to test if the best candidate is yanked, we don't need the full list – only `if getattr(best_candidate, 'yanked', False)`. The list could be made only if the first `best_candidate` is yanked.